api/types/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+            0b11 => Ok(Self::Production),
+            _ => Err(()),
+        }
+    }
+}
+impl From<DeviceLifecycle> for u32 {
+    fn from(value: DeviceLifecycle) -> Self {
+        value as u32
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub struct SecurityState(u32);
+impl From<u32> for SecurityState {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+impl From<SecurityState> for u32 {
+    fn from(value: SecurityState) -> Self {
+        value.0
+    }
+}
+
+impl SecurityState {
+    pub fn debug_locked(self) -> bool {
+        (self.0 & (1 << 2)) != 0
+    }
+    pub fn set_debug_locked(&mut self, val: bool) -> &mut Self {
+        let mask = 1 << 2;
+        if val {
+            self.0 |= mask;
+        } else {
+            self.0 &= !mask
+        };
+        self
+    }
+    pub fn device_lifecycle(self) -> DeviceLifecycle {
+        DeviceLifecycle::try_from(self.0 & 0x3).unwrap()
+    }
+    pub fn set_device_lifecycle(&mut self, val: DeviceLifecycle) -> &mut Self {
+        self.0 |= (val as u32) & 0x3;
+        self
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Default)]
+pub enum U4 {
+    #[default]
+    X0 = 0x0,
+    X1 = 0x1,
+    X2 = 0x2,
+    X3 = 0x3,
+    X4 = 0x4,
+    X5 = 0x5,
+    X6 = 0x6,
+    X7 = 0x7,
+    X8 = 0x8,
+    X9 = 0x9,
+    Xa = 0xa,
+    Xb = 0xb,
+    Xc = 0xc,
+    Xd = 0xd,
+    Xe = 0xe,
+    Xf = 0xf,
+}
+impl U4 {
+    pub const B0000: Self = Self::X0;
+    pub const B0001: Self = Self::X1;
+    pub const B0010: Self = Self::X2;
+    pub const B0011: Self = Self::X3;
+    pub const B0100: Self = Self::X4;
+    pub const B0101: Self = Self::X5;
+    pub const B0110: Self = Self::X6;
+    pub const B0111: Self = Self::X7;
+    pub const B1000: Self = Self::X8;
+    pub const B1001: Self = Self::X9;
+    pub const B1010: Self = Self::Xa;
+    pub const B1011: Self = Self::Xb;
+    pub const B1100: Self = Self::Xc;
+    pub const B1101: Self = Self::Xd;
+    pub const B1110: Self = Self::Xe;
+    pub const B1111: Self = Self::Xf;
+}
+impl From<U4> for u32 {
+    fn from(value: U4) -> Self {
+        value as u32
+    }
+}
+
+impl TryFrom<u32> for U4 {
+    type Error = ();
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            0b0000 => Ok(Self::X0),
+            0b0001 => Ok(Self::X1),
+            0b0010 => Ok(Self::X2),
+            0b0011 => Ok(Self::X3),
+            0b0100 => Ok(Self::X4),
+            0b0101 => Ok(Self::X5),
+            0b0110 => Ok(Self::X6),
+            0b0111 => Ok(Self::X7),
+            0b1000 => Ok(Self::X8),
+            0b1001 => Ok(Self::X9),
+            0b1010 => Ok(Self::Xa),
+            0b1011 => Ok(Self::Xb),
+            0b1100 => Ok(Self::Xc),
+            0b1101 => Ok(Self::Xd),
+            0b1110 => Ok(Self::Xe),
+            0b1111 => Ok(Self::Xf),
+            16_u32..=u32::MAX => Err(()),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
```

Comment:
This feels like a pretty big struct to have `Copy` semantics. It feels like `Clone` would be preferred so that people don't accidentally copy it and not understand the performance penalty.

Reply:
Done in part-2 as well.